### PR TITLE
fix: rebase the hot reload support on webpack deb server

### PR DESCRIPTION
**Pull request by amille44420**
Also fix a critical issue with the manifest file. Whenever there was a change in the initial chuncks, the SSR delivering the entry point of the application would not provide the required bundles. By doing so, webpack remains stuck idling, waiting for the initial bundles to load

We cannot manually remove the require cache because of the way webpack bundle external dependencies on node target. So for dev purposes we use `import-fresh` to ensure we always provide the latest manifest. While on production we switch to a singleton design.

### CODE REVIEW

File: api/index.js [hunk 1]
```diff
@@ -8,8 +8,13 @@ import PrettyError from 'pretty-error';
 import history from 'connect-history-api-fallback';
 import Html from './components/Html';
 import exampleController from './controllers/exampleController';
-// eslint-disable-next-line import/no-unresolved
-import manifest from './chunk-manifest.json';
+
+const getManifest = () => {
+    // eslint-disable-next-line global-require, import/no-extraneous-dependencies
+    const manifestRequire = __DEV__ ? require('import-fresh') : require;
```
Comment (jvasseur) [inline api/index.js]:
```suggestion
    const requireManifest = __DEV__ ? require('import-fresh') : require;
```
?

File: api/index.js [hunk 2]
```diff
@@ -44,14 +49,33 @@ server.use('/example', exampleController);
...
+
+    const html = `<!doctype html>${renderToStaticMarkup(element)}`;
+
+    if (__DEV__) {
+        return html;
+    }
+
+    // on production we don't need to compute it more than once
+    // override the function and use it as a singleton
+    getHtml = () => html;
```
Comment (jvasseur) [inline api/index.js]: I'm not really found of this kind of trick to replace a previously defined function. It makes the code more performant but makes it harder to understand.

File: tools/webpack.config.js [hunk 3]
```diff
@@ -200,6 +213,11 @@ const clientConfig = {
                         for (const chunk of entry.chunks) {
                             // then on files for each chunk
                             for (const file of chunk.files) {
+                                if (file.endsWith('.hot-update.js')) {
+                                    // eslint-disable-next-line no-continue
```
Comment (jvasseur) [inline tools/webpack.config.js]: Maybe we could completely disable this rule in https://github.com/thetribeio/eslint-config ? Personally I think continue has it's use so I would be in favor of removing it.

Comment (amille44420) [reply tools/webpack.config.js]: I agree with your opinion on this rule

File: api/index.js (hunk 2 above)
Comment (amille44420) [reply api/index.js]: It's indeed more tricky but for performances on production it's the best way I'm aware of. I believe the comments we put on it should be enough to explain the trick to anyone reading this line.

**Comment by amille44420** [comment]
I fixed functional tests by setting `disableHostCheck` to `true` on `WebpackDevServer`

File: tools/webpack.config.js (hunk 3 above)
Comment (jvasseur) [reply tools/webpack.config.js]: https://github.com/thetribeio/eslint-config/pull/4

File: api/index.js (hunk 2 above)
Comment (jvasseur) [reply api/index.js]:
```js
const renderHtml = () => {
    // get the manifest (and support async manifest updates on dev mode)
    const { client: manifest } = getManifest();

    // prepare the html for index.html
    const element = createElement(Html, {
        appData,
        manifest,
    });

    return `<!doctype html>${renderToStaticMarkup(element)}`;
};

let getHtml;

if (__DEV__) {
    getHtml = renderHtml;
} else {
    const html = renderHtml();

    getHtml = () => html;
}
```

**Pull request merged**